src/Microsoft.OData.Core/MultipartMixed/ODataMultipartMixedBatchWriter.cs

```diff
@@ -0,0 +1,569 @@
+        }
+
+        /// <summary>
+        /// Verifies that calling CreateOperationRequestMessage is valid.
+        /// </summary>
+        /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
+        /// <param name="method">The Http method to be used for this request operation.</param>
+        /// <param name="uri">The Uri to be used for this request operation.</param>
+        /// <param name="contentId">The Content-ID value to write in ChangeSet head.</param>
+        protected override void VerifyCanCreateOperationRequestMessage(bool synchronousCall, string method, Uri uri, string contentId)
+        {
+            this.CanCreateOperationRequestMessageVerifierCommon(synchronousCall, method, uri, contentId);
+            VerifyCanCreateOperationRequestMessageAgainstChangeSetBoundary(method, contentId);
+        }
+
+        /// <summary>
+        /// Verifies that the writer is in correct state for the Flush operation.
+        /// </summary>
+        /// <param name="synchronousCall">true if the call is to be synchronous; false otherwise.</param>
+        protected override void VerifyCanFlush(bool synchronousCall)
+        {
+            this.RawOutputContext.VerifyNotDisposed();
+            this.VerifyCallAllowed(synchronousCall);
+            if (this.State == BatchWriterState.OperationStreamRequested)
+            {
+                this.ThrowODataException(Strings.ODataBatchWriter_FlushOrFlushAsyncCalledInStreamRequestedState);
+            }
+        }
+
+        /// <summary>
+        /// Sets a new writer state; verifies that the transition from the current state into new state is valid.
+        /// </summary>
+        /// <param name="newState">The writer state to transition into.</param>
+        protected override void SetState(BatchWriterState newState)
+        {
+            this.InterceptException(() => this.ValidateTransition(
+                newState,
+                () => ValidateTransitionAgainstChangesetBoundary(newState, this.changeSetBoundary)));
+
+            switch (newState)
+            {
+                case BatchWriterState.BatchStarted:
+                    Debug.Assert(!this.batchStartBoundaryWritten, "The batch boundary must not be written before calling WriteStartBatch.");
+                    break;
+                case BatchWriterState.ChangesetStarted:
+                    Debug.Assert(this.changeSetBoundary == null, "this.changeSetBoundary == null");
+                    this.changeSetBoundary = ODataBatchWriterUtils.CreateChangeSetBoundary(this.RawOutputContext.WritingResponse);
+                    break;
+                case BatchWriterState.ChangesetCompleted:
+                    Debug.Assert(this.changeSetBoundary != null, "this.changeSetBoundary != null");
+                    this.changeSetBoundary = null;
+                    break;
+            }
+
+            this.State = newState;
+        }

```

Comment:
Setting state should be done in the base class.  Any validation specific to the derived type should be done prior to setting the state.